hail/src/main/scala/is/hail/backend/service/ServiceBackend.scala

```diff
@@ -82,146 +105,206 @@ class ServiceBackend(
-              JString(root),
-              JString(s"$i"))),
-            "type" -> JString("jvm")),
-          "mount_tokens" -> JBool(true))
+        "always_run" -> JBool(false),
+        "job_id" -> JInt(i + 1),
+        "parent_ids" -> JArray(List()),
+        "process" -> JObject(
+          "command" -> JArray(List(
+            JString("is.hail.backend.service.Worker"),
+            JString(revision),
+            JString(jarLocation),
+            JString(root),
+            JString(s"$i"))),
+          "type" -> JString("jvm")),
+        "mount_tokens" -> JBool(true))
       i += 1
     }
 
-    log.info(s"parallelizeAndComputeWithIndex: token $token: running job")
+    log.info(s"parallelizeAndComputeWithIndex: $token: running job")
 
-    val batchClient = BatchClient.fromSessionID(backendContext.sessionID)
-    val batch = batchClient.run(
-      JObject(
-        "billing_project" -> JString(backendContext.billingProject),
-        "n_jobs" -> JInt(n),
-        "token" -> JString(token)),
-      jobs)
+    val batchId = batchClient.create(
+        JObject(
+          "billing_project" -> JString(backendContext.billingProject),
+          "n_jobs" -> JInt(n),
+          "token" -> JString(token),
+          "attributes" -> JObject("name" -> JString(name + "_" + batchCount))),
+        jobs)
+    }
+
+    val batch = batchClient.waitForBatch(batchId)
+    batchCount += 1
     implicit val formats: Formats = DefaultFormats
     val batchID = (batch \ "id").extract[Int]
     val batchState = (batch \ "state").extract[String]
-    if (batchState != "success")
-      throw new RuntimeException(s"batch $batchID failed: $batchState")
+    if (batchState != "success") {
+      throw new HailBatchFailure(s"$batchID")
+    }
 
-    log.info(s"parallelizeAndComputeWithIndex: token $token: reading results")
+    log.info(s"parallelizeAndComputeWithIndex: $token: reading results")
 
     val r = new Array[Array[Byte]](n)
 
-    i = 0  // reusing
-    while (i < n) {
-      r(i) = using(fs.openCachedNoCompression(s"$root/result.$i")) { is =>
-        IOUtils.toByteArray(is)
+    def readResult(i: Int): scalaConcurrent.Future[Unit] = scalaConcurrent.Future {
+      r(i) = retryTransientErrors {
+        using(fs.openCachedNoCompression(s"$root/result.$i")) { is =>
+          IOUtils.toByteArray(is)
+        }
       }
-      i += 1
+      log.info(s"result $i complete")
     }
+
+    scalaConcurrent.Await.result(
+      scalaConcurrent.Future.sequence(
+        Array.tabulate(n)(readResult).toFastIndexedSeq),
+      scalaConcurrent.duration.Duration.Inf)
+
+    log.info(s"all results complete")
     r
   }
 
   def stop(): Unit = ()
 
-  def valueType(username: String, s: String): String = {
-    ExecutionTimer.logTime("ServiceBackend.valueType") { timer =>
-      userContext(username, timer) { ctx =>
-        val x = IRParser.parse_value_ir(ctx, s)
-        x.typ.toString
-      }
-    }
+  def valueType(
```

Comment:
I wonder if should support only a single execute endpoint by adding functions / ir nodes to look up the types and return them as strings. Might be a bit simpler. Maybe not.

Reply:
I am game for whatever API you all like. The current one is not particularly onerous for me to support.